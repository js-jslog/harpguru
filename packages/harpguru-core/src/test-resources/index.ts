export {
  inactiveCellsHarpStrata,
  activeCellsHarpStrata,
  buildMockUseGlobalImplementation,
} from './test-resources'
