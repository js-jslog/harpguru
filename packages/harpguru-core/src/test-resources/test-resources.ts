import { getHarpStrata } from 'harpstrata'
import type { HarpStrata, HarpStrataProps, ActiveIds } from 'harpstrata'
import {
  TuningIds,
  PitchIds,
  DegreeIds,
  PozitionIds,
  ValvingIds,
} from 'harpparts'

import {
  reduceFullMatrixToViewableMatrix,
  reduceViewableMatrixToLayoutFacts,
} from '../utils'
import { DisplayModes, ExperienceModes, FlushChannels } from '../types'
import type { ColumnBounds } from '../types'

const allActiveDegrees = [
  DegreeIds.Root,
  DegreeIds.Flat2,
  DegreeIds.Second,
  DegreeIds.Flat3,
  DegreeIds.Third,
  DegreeIds.Fourth,
  DegreeIds.Flat5,
  DegreeIds.Fifth,
  DegreeIds.Flat6,
  DegreeIds.Sixth,
  DegreeIds.Flat7,
  DegreeIds.Seventh,
]

const baseHarpStrataProps: HarpStrataProps = {
  tuningId: TuningIds.MajorDiatonic,
  valvingId: ValvingIds.NotValved,
  pozitionId: PozitionIds.First,
  harpKeyId: PitchIds.C,
  activeIds: [] as ActiveIds,
}
const inactiveCellsHarpStrataProps = baseHarpStrataProps
const activeCellsHarpStrataProps = {
  ...baseHarpStrataProps,
  activeIds: allActiveDegrees,
}

export const inactiveCellsHarpStrata = getHarpStrata(
  inactiveCellsHarpStrataProps
)
export const activeCellsHarpStrata = getHarpStrata(activeCellsHarpStrataProps)

type BuildMockUseGlobalImplementationProps = {
  readonly sourceHarpStrata?: HarpStrata
  readonly sourceColumnBounds?: ColumnBounds
  readonly experienceMode?: ExperienceModes
  readonly displayMode?: DisplayModes
  readonly bufferedActivityToggles?: ReadonlyArray<DegreeIds>
  readonly fragmentHarpFaceByOctaves?: boolean
  readonly flushChannel?: FlushChannels
  readonly activeQuizDegrees?: ReadonlyArray<DegreeIds>
}

type MockGlobalImplementation = (arg0: string) => unknown[] | undefined

export const buildMockUseGlobalImplementation = ({
  sourceHarpStrata = activeCellsHarpStrata,
  sourceColumnBounds = 'FIT',
  experienceMode = ExperienceModes.Explore,
  displayMode = DisplayModes.Degree,
  bufferedActivityToggles = [],
  fragmentHarpFaceByOctaves = true,
  flushChannel = FlushChannels.Regular,
  activeQuizDegrees = allActiveDegrees,
}: BuildMockUseGlobalImplementationProps): MockGlobalImplementation => {
  const mockGlobal = (stateItem: string) => {
    if (stateItem === 'activeDegreeMatrix')
      return [sourceHarpStrata.degreeMatrix]
    if (stateItem === 'activePitchMatrix') return [sourceHarpStrata.pitchMatrix]
    if (stateItem === 'activeInteractionMatrix')
      return [sourceHarpStrata.apparatus.interactionMatrix]
    if (stateItem === 'activeDegreeIds')
      return [sourceHarpStrata.activeDegreeIds]
    if (stateItem === 'activePitchIds') return [sourceHarpStrata.activePitchIds]
    if (stateItem === 'tuningId') return [sourceHarpStrata.apparatus.tuningId]
    if (stateItem === 'valvingId') return [sourceHarpStrata.apparatus.valvingId]
    if (stateItem === 'harpKeyId') return [sourceHarpStrata.harpKeyId]
    if (stateItem === 'pozitionId') return [sourceHarpStrata.pozitionId]
    if (stateItem === 'rootPitchId') return [sourceHarpStrata.rootPitchId]

    if (stateItem === 'columnBounds') return [sourceColumnBounds]
    const viewableDegreeMatrix = reduceFullMatrixToViewableMatrix(
      [[]] as const,
      sourceHarpStrata.degreeMatrix,
      sourceColumnBounds
    )
    const viewablePitchMatrix = reduceFullMatrixToViewableMatrix(
      [[]] as const,
      sourceHarpStrata.pitchMatrix,
      sourceColumnBounds
    )
    const viewableInteractionMatrix = reduceFullMatrixToViewableMatrix(
      [[]] as const,
      sourceHarpStrata.apparatus.interactionMatrix,
      sourceColumnBounds
    )
    if (stateItem === 'viewableDegreeMatrix') return [viewableDegreeMatrix]
    if (stateItem === 'viewablePitchMatrix') return [viewablePitchMatrix]
    if (stateItem === 'viewableInteractionMatrix')
      return [viewableInteractionMatrix]

    const layoutFacts = reduceViewableMatrixToLayoutFacts(
      viewableInteractionMatrix
    )
    if (stateItem === 'layoutFacts') return [layoutFacts]

    if (stateItem === 'activeExperienceMode') return [experienceMode]
    if (stateItem === 'activeDisplayMode') return [displayMode]
    if (stateItem === 'bufferedActivityToggles')
      return [bufferedActivityToggles]
    if (stateItem === 'fragmentHarpFaceByOctaves')
      return [fragmentHarpFaceByOctaves]
    if (stateItem === 'flushChannel') return [flushChannel]
    if (stateItem === 'activeQuizDegrees') return [activeQuizDegrees]
    return undefined
  }
  return mockGlobal
}
