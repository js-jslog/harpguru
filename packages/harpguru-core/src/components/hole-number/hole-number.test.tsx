import { useGlobal } from 'reactn'
import React from 'react'
import { render } from '@testing-library/react-native'

import { activeCellsHarpStrata } from '../../test-resources'

import { HoleNumber, HoleNumberIds } from './hole-number'

jest.mock('reactn')
const mockUseGlobal = useGlobal as jest.Mock
mockUseGlobal.mockImplementation((stateItem: string) => {
  if (stateItem === 'activeHarpStrata') return [activeCellsHarpStrata]
  return undefined
})

test('HoleNumber renders a dom element with the expected value included', () => {
  const { getByText } = render(<HoleNumber xCoord={0} />)
  expect(getByText(HoleNumberIds.One)).toBeTruthy()
})

test('A snapshot of HoleNumber', () => {
  const { toJSON } = render(<HoleNumber xCoord={0} />)
  expect(toJSON()).toMatchSnapshot()
})
