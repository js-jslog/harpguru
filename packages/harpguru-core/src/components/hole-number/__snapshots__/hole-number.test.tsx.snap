// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A snapshot of HoleNumber 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "height": 0,
        "justifyContent": "center",
        "width": 59.270470780512596,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#555",
          "fontSize": 22.63930291919421,
        }
      }
    >
      1
    </Text>
  </View>
</View>
`;
